### PR TITLE
fix: Track tax switch does not appear grayed out when enabled/disabled offline

### PR DESCRIPTION
**Pull request by nkdengineer**
### Fixed Issues

$ https://github.com/Expensify/App/issues/43318
PROPOSAL:

### Tests

1. Go to workspace settings > Distance rates
2. Go offline
3. Click Settings
4. Enable or disable Track tax switch
5. Verify that: Track tax switch will appear grayed out when enabled/disabled offline

- [x] Verify that no errors appear in the JS console

### QA Steps

1. Go to workspace settings > Distance rates
2. Go offline
3. Click Settings
4. Enable or disable Track tax switch
5. Verify that: Track tax switch will appear grayed out when enabled/disabled offline

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

android

</details>

<details>
<summary>Android: mWeb Chrome</summary>

android-mweb

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/161821005/3ab7e403-05cf-4fe9-a9eb-3bc14eefa9b6

</details>

<details>
<summary>iOS: mWeb Safari</summary>

ios-mweb

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/161821005/821c7f5f-f889-4631-9a3a-90917466ecd8

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/161821005/cb2036da-4f89-46e6-929a-909383c157f8

</details>

### CODE REVIEW

**Comment by nkdengineer** [comment]
@eVoloshchak Please help to review the PR when you have a chance. Thanks.

**Comment by srikarparsi** [comment]
@eVoloshchak please tag me after you review

File: src/types/onyx/Policy.ts [hunk 1]
```diff
@@ -57,31 +57,34 @@ type Attributes = {
 };
 
 /** Policy custom unit */
-type CustomUnit = OnyxCommon.OnyxValueWithOfflineFeedback<{
-    /** Custom unit name */
-    name: string;
+type CustomUnit = OnyxCommon.OnyxValueWithOfflineFeedback<
```
Comment (eVoloshchak) [inline src/types/onyx/Policy.ts:60]: Was this change intended?

**Comment by eVoloshchak** [comment]
@nkdengineer, could you please add the screen recording for all of the platforms?
(you can four-finger tap to open a menu with `Force Offline` option)

File: src/types/onyx/Policy.ts (hunk 1 above)
Comment (nkdengineer) [reply src/types/onyx/Policy.ts:60]: Yes, we added `keyof Attributes` at the end so it'll have this format

**Comment by nkdengineer** [comment]
@eVoloshchak please check again

**Comment by srikarparsi** [comment]
Hey @eVoloshchak can you take a look at this please?

**Comment by nkdengineer** [comment]
friendly dump @eVoloshchak

**Comment by eVoloshchak** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

**Review by eVoloshchak — COMMENTED** [review]
**App crash:**

1. Clear cache
2. Force offline
3. Go to workspace settings (has to be a workspace you haven't opened yet, no workspace data is loaded at this point)
4. Go to 'Distance rates' tab
5. Click Settings
6. The app crashes


https://github.com/Expensify/App/assets/9059945/b3b3dd0b-135b-4fbd-b262-7839139bb2aa

**Comment by nkdengineer** [comment]
I can't reproduce this bug
After `Clear cache` it still has workspace data

https://github.com/Expensify/App/assets/161821005/cc09ac75-886e-4b84-8984-c7f836fd9c67

**Review by eVoloshchak — COMMENTED** [review]
@nkdengineer, you can more reliably reproduce this using the following steps:

1. Fully clear cookies and site data in browser (`View site information -> Cookies and site data -> Manage on-device site data -> Delete each item -> Done`)
2. Refresh the page
3. Log in
4. Go to `Settings -> Troubleshoot`, turn on `Force Offline` option
5. Head to Workspaces, select any workspace
6. Select Distance Rates -> Settings
7. Observe the crash


https://github.com/Expensify/App/assets/9059945/f00fe3c2-1b1c-4ae7-a187-5db737db96d0

**Comment by eVoloshchak** [comment]
**Bug:** 
1. Go to `Settings -> Workspaces`
2. Create a new workspace
3. Go to More Features
4. Turn on `Distance Rates` and `Taxes`
5. Head to `Distance Rates -> Settings`
6. `Track Tax` switch is not responsive
7. The error is displayed when changing `Unit` and `Default Category`

https://github.com/Expensify/App/assets/9059945/a5f0cb06-ccb2-49df-ae06-3477945b2b9b

**Comment by nkdengineer** [comment]
Working on this issue

**Comment by nkdengineer** [comment]
@eVoloshchak thanks for your detailed steps, I can reproduce this bug

https://github.com/Expensify/App/assets/161821005/222673c7-c498-4134-84c9-f4a07b9e3d35


in this case, we don't have `customUnits` data, I think we should block the page offline if data is already available
like we do with tag settings page in this [PR](https://github.com/Expensify/App/pull/42623)

**Comment by nkdengineer** [comment]
@eVoloshchak i think this [bug](https://github.com/Expensify/App/pull/43347#issuecomment-2192197140) is a backend issue.
We created a new workspace with the parameter `customUnitID = 'BDF9F0A7F310D'`, but the backend returned a new `customUnitID = 'A4498E5F3FE8F'`. Consequently, we encountered an error `Invalid ID in custom unit` when attempting to update some fields.

https://github.com/Expensify/App/assets/161821005/e2baa845-2a7f-4762-98ee-da00220ca111

**Comment by nkdengineer** [comment]
@eVoloshchak friendly bump

**Comment by eVoloshchak** [comment]
>  i think this https://github.com/Expensify/App/pull/43347#issuecomment-2192197140 is a backend issue.

I'm unable to reproduce this, looks like this was resolved on BE

> in this case, we don't have customUnits data, I think we should block the page in offline if data is already available
like we do with tag settings page in this https://github.com/Expensify/App/pull/42623

@nkdengineer, could we simply show a loading spinner while the data is loading?

**Comment by nkdengineer** [comment]
> could we simply show a loading spinner while the data is loading?

@eVoloshchak In offline mode, we can't get the data and will get an infinite loading spinner

**Comment by nkdengineer** [comment]
@eVoloshchak friendly bump

**Comment by eVoloshchak** [comment]
@nkdengineer, in that case, let's disable the settings button if we're offline and there are no distance rates available. But there still will be a case when the app is online, and the distance rates haven't loaded yet (currently it just crashes in this case), we need to account for that too

**Comment by trjExpensify** [comment]
@MonilBhavsar can you have eyes on this PR, please? 👀

**Comment by MonilBhavsar** [comment]
> I'm unable to reproduce this, looks like this was resolved on BE

Looks like the issue being discussed was resolved and now we've a different issue.
Can anyone please explain the issue. Looks like in offline mode, the tax enable switch is not working as expected and crashing the app?

**Comment by nkdengineer** [comment]
@MonilBhavsar please check this [comment](https://github.com/Expensify/App/pull/43347#issuecomment-2196333838)

**Comment by nkdengineer** [comment]
@eVoloshchak @MonilBhavsar i fixed this [bug](https://github.com/Expensify/App/pull/43347#pullrequestreview-2142450762)
in case of isOffline and we don't have data, we'll show the offline page like this 

Screenshot 2024-07-10 at 23 02 10

and in case of isOnline and we don't have data, it's working now


https://github.com/Expensify/App/assets/161821005/a2ec8787-9015-421b-aaa4-89522c63e60b

**Comment by trjExpensify** [comment]
Is this ready for re-review by @eVoloshchak & @MonilBhavsar?

File: src/pages/workspace/distanceRates/PolicyDistanceRatesSettingsPage.tsx [hunk 2]
```diff
@@ -91,68 +164,7 @@ function PolicyDistanceRatesSettingsPage({policy, policyCategories, route}: Poli
...
-                                            {translate('workspace.common.moreFeatures')}
-                                        </TextLink>
-                                        {translate('workspace.distanceRates.changePromptMessage')}
-                                    </Text>
-                                </View>
-                            )}
-                        </OfflineWithFeedback>
-                    </View>
-                </ScrollView>
+                {!customUnit && isOffline ? <FullPageOfflineBlockingView>{getDistanceRatesSetting()}</FullPageOfflineBlockingView> : getDistanceRatesSetting()}
```
[the remark this answers is not in the document]
Comment (nkdengineer) [reply src/pages/workspace/distanceRates/PolicyDistanceRatesSettingsPage.tsx]: @eVoloshchak i updated

**Review by eVoloshchak — COMMENTED** [review]
@nkdengineer, the page crashes when connection is restored 

https://github.com/user-attachments/assets/b08b926a-dc64-44eb-8f3a-68c7c50d4b1a

**Comment by trjExpensify** [comment]
@nkdengineer can you address the comments?

**Comment by nkdengineer** [comment]
@eVoloshchak I fixed this bug. Please check again.

File: src/pages/workspace/distanceRates/UnitSelector/index.tsx [hunk 3]
```diff
@@ -41,7 +41,7 @@ function UnitSelector({defaultValue, wrapperStyle, label, setNewUnit}: UnitSelec
         hidePickerModal();
     };
 
-    const title = Str.recapitalize(translate(getUnitTranslationKey(defaultValue)));
+    const title = defaultValue ? Str.recapitalize(translate(getUnitTranslationKey(defaultValue))) : '';
```
Comment (eVoloshchak) [inline src/pages/workspace/distanceRates/UnitSelector/index.tsx:44]: If `defaultValue` can be `undefined`, then we need to mark it as optional in `UnitSelectorProps`

**Comment by eVoloshchak** [comment]
Bug:  distance rates list flashes every time you press `Settings` or `Add Rate` button


https://github.com/user-attachments/assets/08697f6b-bfe2-40bf-97e7-60b759c38b7b

**Comment by nkdengineer** [comment]
I can't reproduce this bug.

https://github.com/user-attachments/assets/5535bc05-9156-4d26-a609-f8ea83ac2815

**Comment by eVoloshchak** [comment]
> I can't reproduce this bug.

@nkdengineer, I'm still trying to pinpoint the exact steps that are causing this bug, meanwhile found another one:

1. Log out, clear cookies and cache
2. Log in
3. Turn off the internet connection
4. Go to Settings -> Workspaces -> Select any workspace
5. Go to Distance Rates
6. Add a new rate
7. Press on the Settings button
8. The app crashes

https://github.com/user-attachments/assets/38b7c5a0-cd37-48ce-b074-134285143bc3

**Comment by nkdengineer** [comment]
@eVoloshchak i fixed this [bug](https://github.com/Expensify/App/pull/43347#issuecomment-2242688133)

**Comment by MonilBhavsar** [comment]
@eVoloshchak would you be able to confirm if the bug has been fixed, please

**Comment by eVoloshchak** [comment]
Can confirm the [bug](https://github.com/Expensify/App/pull/43347#issuecomment-2242688133) was fixed, but found a couple of new ones for the tax rates created while offline:


1. Log out, clear cookies and cache
2. Log in
3. Turn off the internet connection
4. Go to Settings -> Workspaces -> Select any workspace
5. Go to Distance Rates
6. Add a distance rate, press on it to edit it's settings
7. Notice you cannot change the rate
8. Notice `Tax rate` and T`ax reclaimable on` menu items are not populated
9. You cannot change `Tax rate` and `Tax reclaimable on` settings
10. Turn the internet connection back on
11. Notice an error appears below the tax rate created while offline
12. Discard the error, reload the page, try to delete the workspace
13. Notice you cannot delete the workspace

All of these issues are reproducible only for the tax rates that were created while offline, I suspect they share a common cause (invalid optimistic ID perhaps?)


https://github.com/user-attachments/assets/40a32c20-6849-405b-996b-50b53ac3bf08

**Comment by nkdengineer** [comment]
I'll update today

**Comment by nkdengineer** [comment]
@MonilBhavsar @eVoloshchak When user clears cache and restart, we call API openApp. You can see in this video below, that we have some workspaces that will not have a default rate 

https://github.com/user-attachments/assets/67ff4f10-2b48-4c97-89cb-4748505bd6bf

It will cause some errors when we create a new distance rate because we don’t have `customUnitsID`

Screenshot 2024-07-29 at 12 12 55

**Comment by MonilBhavsar** [comment]
You think that not having custom unit ID is wrong and we should fix this case in backend?

**Comment by nkdengineer** [comment]
@MonilBhavsar If the backend can return `customUnitID` of all workspaces when calling `OpenApp` API, we will not encounter bugs like the ones mentioned above.

**Comment by MonilBhavsar** [comment]
I see. AFAIK, We decided not to load all policies on opening app for performance reasons. Basically to reduce initial app loading time. So we need to be optimistic here. Can we load it when we open a specific workspace page?

**Comment by nkdengineer** [comment]
Let me check again if we can have a solution in front-end.

**Comment by nkdengineer** [comment]
@MonilBhavsar @eVoloshchak
When offline and the workspace does not have a customUnit, we will not know the `customUnitID` and will send `customUnitID = ''`. This will result in an error as follows:

bug

I think in the case when offline and there is no customUnit, we should not allow adding a new rate to display the `FullPageOfflineBlockingView` page.

**Comment by nkdengineer** [comment]
I'll update today

**Comment by nkdengineer** [comment]
@eVoloshchak i updated, please check agin

https://github.com/user-attachments/assets/56e23c22-1efb-4d6b-8e37-375c2304fb43

**Review by eVoloshchak — APPROVED** [review]
LGTM!

File: src/pages/workspace/distanceRates/UnitSelector/index.tsx (hunk 3 above)
Comment (srikarparsi) [reply src/pages/workspace/distanceRates/UnitSelector/index.tsx:44]: I agree, lets make this field optional

**Review by srikarparsi — DISMISSED** [review]
This looks good to me other than making `defaultValue` optional. Testing between offline and online mode also works well. Thank you @nkdengineer and @eVoloshchak for covering the edge cases.

**Comment by srikarparsi** [comment]
Also thank you @MonilBhavsar for helping move this along. I added you to the initial [issue](https://github.com/Expensify/App/issues/43318).

File: src/pages/workspace/distanceRates/UnitSelector/index.tsx (hunk 3 above)
Comment (MonilBhavsar) [reply src/pages/workspace/distanceRates/UnitSelector/index.tsx:44]: @nkdengineer ^

Comment (nkdengineer) [reply src/pages/workspace/distanceRates/UnitSelector/index.tsx:44]: @MonilBhavsar @srikarparsi  I updated.

**Review by MonilBhavsar — APPROVED** [review]
Looks good

**Comment by MonilBhavsar** [comment]
Thanks for continuing working on this!

**Comment by Beamanator** [comment]
FYI I believe this was deployed to prod yesterday, from this checklist - #47356
